File: maxHeap.java
import java.util.*;

public class maxHeap {
    private ArrayList<Barang> heap = new ArrayList<>();

    
    public void insert(Barang data) {
        heap.add(data);
        int index = heap.size() - 1;

        
        while (index > 0 && heap.get((index - 1) / 2).id < heap.get(index).id) {
            Collections.swap(heap, index, (index - 1) / 2);
            index = (index - 1) / 2;
        }
    }

    
    public Barang extractMax() {
        if (heap.isEmpty()) return null;
        
        Barang root = heap.get(0);
        Barang lastItem = heap.remove(heap.size() - 1);
        
        if (!heap.isEmpty()) {
            heap.set(0, lastItem);
            maxHeapify(0);
        }
        return root;
    }

    private void maxHeapify(int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < heap.size() && heap.get(l).id > heap.get(largest).id) largest = l;
        if (r < heap.size() && heap.get(r).id > heap.get(largest).id) largest = r;

        if (largest != i) {
            Collections.swap(heap, i, largest);
            maxHeapify(largest);
        }
    }

    public boolean isEmpty() {
        return heap.isEmpty();
    }

    
    public ArrayList<Barang> getHeapData() {
        return new ArrayList<>(heap);
    }
}

File: maxheap.py
def insert(heap, data):
    """
    data adalah tuple: (id, nama)
    Kita bandingkan berdasarkan data[0] yaitu ID-nya.
    """
    heap.append(data)
    index = len(heap) - 1

    
    while index > 0 and heap[(index - 1) // 2][0] < heap[index][0]:
        heap[index], heap[(index - 1) // 2] = heap[(index - 1) // 2], heap[index]
        index = (index - 1) // 2

def extract_max(heap):
    """
    Menghapus dan mengembalikan elemen dengan ID terbesar (root).
    """
    if not heap:
        return None
    if len(heap) == 1:
        return heap.pop()
    
    root = heap[0]
    heap[0] = heap.pop() 
    _max_heapify(heap, 0)
    return root

def _max_heapify(heap, i):
    """
    Menyeimbangkan kembali heap setelah root dihapus.
    """
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2
    n = len(heap)

    if l < n and heap[l][0] > heap[largest][0]:
        largest = l
    if r < n and heap[r][0] > heap[largest][0]:
        largest = r

    if largest != i:
        heap[i], heap[largest] = heap[largest], heap[i]
        _max_heapify(heap, largest)


if __name__ == "__main__":
    arr = []
    
    values = [(10, "Buku"), (7, "Pena"), (11, "Meja"), (5, "Kursi"), (13, "Lampu")]

    print("--- Proses Insert ke Max-Heap ---")
    for val in values:
        insert(arr, val)
        print(f"Inserted {val} -> Current Heap: {arr}")

    print("\n--- Proses Hapus (Extract Max) ---")
    while arr:
        removed = extract_max(arr)
        print(f"Dihapus (Terbesar): {removed} | Sisa Heap: {arr}")